src/Symfony/Component/CssSelector/XPath/Extension/RelationExtension.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\CssSelector\XPath\Extension;
+
+use Symfony\Component\CssSelector\XPath\XPathExpr;
+
+/**
+ * XPath expression translator combination extension.
+ *
+ * This component is a port of the Python cssselect library,
+ * which is copyright Ian Bicking, @see https://github.com/scrapy/cssselect.
+ *
+ * @author Franck Ranaivo-Harisoa <franckranaivo@gmail.com>
+ *
+ * @internal
+ */
+class RelationExtension extends AbstractExtension
+{
+    public function getRelativeCombinationTranslators(): array
+    {
+        return [
+            ' ' => $this->translateRelationDescendant(...),
+            '>' => $this->translateRelationChild(...),
+            '+' => $this->translateRelationDirectAdjacent(...),
+            '~' => $this->translateRelationIndirectAdjacent(...),
+        ];
+    }
+
+    public function translateRelationDescendant(XPathExpr $xpath, XPathExpr $combinedXpath): XPathExpr
+    {
+        return $xpath->join('[descendant-or-self::', $combinedXpath, ']', true);
+    }
+
+    public function translateRelationChild(XPathExpr $xpath, XPathExpr $combinedXpath): XPathExpr
+    {
+        return $xpath->join('[./', $combinedXpath, ']');
+    }
+
+    public function translateRelationDirectAdjacent(XPathExpr $xpath, XPathExpr $combinedXpath): XPathExpr
+    {
+        return $xpath
+            ->addCondition(sprintf('following-sibling::*[(name() = \'%s\') and (position() = 1)]', $combinedXpath->getElement()));
```

Comment:
using only `$combinedXpath->getElement()` looks suspicious to me. What if the selector is `:has(+ .baz)` ?

Reply:
Ah yes, you're right, but python code only handle `->getElement()` .  as you can see [here](https://github.com/scrapy/cssselect/blob/ddd9784977fca6e7da4439d837e4f510f1f10638/cssselect/xpath.py#L445) .

I will try if I can find a better alternative elsewhere.